pkg/util/winutil/eventlog/subscription/subscription.go

```diff
@@ -0,0 +1,366 @@
+		q.query,
+		bookmarkHandle,
+		q.subscribeOriginFlag|q.subscribeFlags)
+	if err != nil {
+		safeCloseNullHandle(windows.Handle(hSubWait))
+		return err
+	}
+
+	// alloc reusable storage for EvtNext output
+	q.evtNextStorage = make([]evtapi.EventRecordHandle, q.eventBatchCount)
+
+	q.subscriptionSignalEventHandle = hSubWait
+	q.subscriptionHandle = hSub
+
+	hStopWait, err := newStopWaitEvent()
+	if err != nil {
+		evtapi.EvtCloseResultSet(q.eventLogAPI, q.subscriptionHandle)
+		safeCloseNullHandle(windows.Handle(hSubWait))
+		return err
+	}
+	q.stopEventHandle = hStopWait
+
+	q.eventsChannel = make(chan []*evtapi.EventRecord)
+
+	q.notifyStop = make(chan struct{})
+
+	// start goroutine to query events for channel
+	q.getEventsLoopWaiter.Add(1)
+	go q.getEventsLoop()
+
+	q.started = true
+
+	return nil
+}
+
+func (q *pullSubscription) Stop() {
+	if !q.started {
+		return
+	}
+
+	// Signal getEventsLoop to stop
+	windows.SetEvent(windows.Handle(q.stopEventHandle))
+	close(q.notifyStop)
+
+	// Wait for getEventsLoop to stop
+	q.getEventsLoopWaiter.Wait()
+
+	safeCloseNullHandle(windows.Handle(q.stopEventHandle))
+
+	// Cleanup Windows API
+	evtapi.EvtCloseResultSet(q.eventLogAPI, q.subscriptionHandle)
+	safeCloseNullHandle(windows.Handle(q.subscriptionSignalEventHandle))
+
+	q.started = false
+}
+
+func (q *pullSubscription) logAndSetError(err error) {
+	pkglog.Error(err)
+	q.err = err
+}
+
+// getEventsLoop waits for events to be available and writes them to eventsChannel.
+// On return, closes the eventsChannel channel to notify the user of an error or a Stop().
+func (q *pullSubscription) getEventsLoop() {
+	// q.Stop() waits on this goroutine to finish, notify it that we are done
+	defer q.getEventsLoopWaiter.Done()
+	defer close(q.eventsChannel)
+
+	waiters := []windows.Handle{windows.Handle(q.subscriptionSignalEventHandle), windows.Handle(q.stopEventHandle)}
+
+	for {
+		dwWait, err := windows.WaitForMultipleObjects(waiters, false, windows.INFINITE)
+		if err != nil {
+			// WAIT_FAILED
+			q.logAndSetError(fmt.Errorf("WaitForMultipleObjects failed: %w", err))
+			return
+		}
+
+		if dwWait == windows.WAIT_OBJECT_0 {
+			// We were signalled by the subscription, check for events
+			pkglog.Trace("Checking for events")
+			// We supply INFINITE for the Timeout parameter but if we are at the end of the log file/there are no more events EvtNext will return,
+			// it will not block forever.
+			// https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-even6/cd4c258c-5a2c-4ba8-bce3-37eefaa416e7
+			eventRecordHandles, err := q.eventLogAPI.EvtNext(q.subscriptionHandle, q.evtNextStorage, uint(len(q.evtNextStorage)), windows.INFINITE)
+			if err == nil {
+				pkglog.Debugf("EvtNext returned %v handles", len(eventRecordHandles))
+				select {
+				case q.eventsChannel <- q.parseEventRecordHandles(eventRecordHandles):
+				case <-q.notifyStop:
+					q.err = fmt.Errorf("received stop signal")
+					pkglog.Info(q.err)
+					return
+				}
```

Comment:
writing to `q.eventsChannel` is a blocking operation, it blocks until someone reads on the channel. So it will block forever unless we also wait on the stop channel.

Reply:
Of course, I thought maybe there is another primitive to "slide" execution lower if it is blocked. But my comment is cosmetic anyway.